### PR TITLE
connectors-ci: centralize patches in `hacks.py`

### PR DESCRIPTION
**Pull request by alafanechere**
## What
For better tech debt mgmt I'm centralizing the "in memory" source code patch we're doing in the CI pipeline to work around certain problems.

### CODE REVIEW

File: tools/ci_connector_ops/ci_connector_ops/pipelines/contexts.py [hunk 1]
```diff
@@ -403,7 +404,8 @@ def get_connector_dir(self, exclude=None, include=None) -> Directory:
         Returns:
             Directory: The connector under test source code directory.
         """
-        return self.get_repo_dir(str(self.connector.code_directory), exclude=exclude, include=include)
+        vanilla_connector_dir = self.get_repo_dir(str(self.connector.code_directory), exclude=exclude, include=include)
+        return await hacks.patch_connector_dir(self, vanilla_connector_dir)
```
Comment (bnchrch) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/contexts.py:408]: hacks! <3

File: tools/ci_connector_ops/ci_connector_ops/pipelines/hacks.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    context.logger.warn("Patching dockerfile to pin Cython to <3.0 and pyyaml to ~=5.4")
+    patched_dockerfile = []
+    pinned = False
+    for line in dockerfile_content.splitlines():
+        # Workaround for https://github.com/yaml/pyyaml/issues/601
+        # Cython 3.0 is incompatible with pyyaml ~=5.4, we shall pin Cython to <3.0
+        # Pyyaml should release a new patch version that explicitly pins Cython to <3.0
+        if "RUN pip install" in line and not pinned:
+            patched_dockerfile.append('RUN pip install --prefix=/install "Cython<3.0" "pyyaml~=5.4" --no-build-isolation')
```
Comment (bnchrch) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/hacks.py:49]: Lets explain why we do append both in the dockerfile.

File: tools/ci_connector_ops/ci_connector_ops/pipelines/hacks.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            patched_dockerfile.append('RUN pip install "Cython<3.0" "pyyaml~=5.4" --no-build-isolation')
+            pinned = True
```
Comment (bnchrch) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/hacks.py:51]: Lets emit a log warning when we do this.

If we forget to remove this, we could accidentally cause a bug confusing the heck out of a dev

**Review by bnchrch — APPROVED** [review]
Fundementally, I feel awkward editing dockerfiles programmatically. But understand we do need to do this to unblock our pipeline.

Before we merge can we confirm this works by prepublishing a connector from this branch?

**Comment by alafanechere** [comment]
@bnchrch I tried out a pre-release here: https://github.com/airbytehq/airbyte/actions/runs/5591968135/jobs/10223781377
It's hanging so far 😢

File: tools/ci_connector_ops/ci_connector_ops/pipelines/hacks.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    context.logger.info("Patching acceptance-test-config.yml to use connector image with git revision tag and not dev.")
+
+    patched_cat_config = deepcopy(context.connector.acceptance_test_config)
+    patched_cat_config["connector_image"] = context.connector.acceptance_test_config["connector_image"].replace(
+        ":dev", f":{context.git_revision}"
+    )
+    return connector_dir.with_new_file("acceptance-test-config.yml", yaml.safe_dump(patched_cat_config))
+
+
+async def patch_connector_dir(context: ConnectorContext, connector_dir: Directory) -> Directory:
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply tools/ci_connector_ops/ci_connector_ops/pipelines/hacks.py:107]: Note that the example uses `patch_connector_dir`

But the bulk of the info likely live in the docstring of the function which applies the hack like `_patch_dockerfile`

**Comment by octavia-squidington-iii** [comment]
##  source-openweather test report (commit [`b5d2150a93`](https://github.com/airbytehq/airbyte/pull/28390/commits/b5d2150a939a8859df25da0c3bb625683f4dd72b)) - ✅

⏲️  Total pipeline duration: 01mn41s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-openweather/metadata.yaml | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Code format checks                                                        | ✅        |
| Connector package install                                                 | ✅        |
| Build source-openweather docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                | ✅        |
| Acceptance tests                                                          | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_patch-python-dockerfiles/1689858842/b5d2150a939a8859df25da0c3bb625683f4dd72b/source-openweather/0.1.6/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-openweather test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`b5d2150a93`](https://github.com/airbytehq/airbyte/pull/28390/commits/b5d2150a939a8859df25da0c3bb625683f4dd72b)) - ❌

⏲️  Total pipeline duration: 04mn29s 

| Step                                                                      | Result   |
|:--------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml     | ✅        |
| Connector version semver check                                            | ✅        |
| QA checks                                                                 | ✅        |
| Build connector tar                                                       | ✅        |
| Build source-alloydb docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-alloydb:integrationTest | ✅        |
| Acceptance tests                                                          | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_patch-python-dockerfiles/1689858842/b5d2150a939a8859df25da0c3bb625683f4dd72b/source-alloydb/2.0.28/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb-strict-encrypt test report (commit [`b5d2150a93`](https://github.com/airbytehq/airbyte/pull/28390/commits/b5d2150a939a8859df25da0c3bb625683f4dd72b)) - ❌

⏲️  Total pipeline duration: 04mn06s 

| Step                                                                                     | Result   |
|:-----------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb-strict-encrypt/metadata.yaml     | ✅        |
| Connector version semver check                                                           | ✅        |
| QA checks                                                                                | ✅        |
| Build connector tar                                                                      | ✅        |
| Build source-alloydb-strict-encrypt docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-alloydb-strict-encrypt:integrationTest | ✅        |
| Acceptance tests                                                                         | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_patch-python-dockerfiles/1689858842/b5d2150a939a8859df25da0c3bb625683f4dd72b/source-alloydb-strict-encrypt/2.0.28/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  destination-bigquery test report (commit [`641ed8ac79`](https://github.com/airbytehq/airbyte/pull/28390/commits/641ed8ac79544538689a14c3758e7e848efe2e6d)) - ❌

⏲️  Total pipeline duration: 28mn06s 

| Step                                                                            | Result   |
|:--------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/destination-bigquery/metadata.yaml     | ✅        |
| Connector version semver check                                                  | ✅        |
| QA checks                                                                       | ✅        |
| Build connector tar                                                             | ✅        |
| Build destination-bigquery docker image for platform linux/x86_64               | ✅        |
| Build airbyte/normalization:dev                                                 | ✅        |
| ./gradlew :airbyte-integrations:connectors:destination-bigquery:integrationTest | ❌        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/augustin_connectors-ci_patch-python-dockerfiles/1689859581/641ed8ac79544538689a14c3758e7e848efe2e6d/destination-bigquery/1.5.3/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=destination-bigquery test
```

**Pull request merged**